### PR TITLE
ref(hc): More test passing in split db mode

### PR DESCRIPTION
**Pull request by corps**
Also fixes obnoxious silo mode decorator bug having to do with `override_settings` and the lifecycle of django applications in unit tests.

### CODE REVIEW

File: src/sentry/db/postgres/transactions.py [hunk 1]
```diff
@@ -33,10 +33,11 @@ def django_test_transaction_water_mark(using: str | None = None):
         return
 
     from sentry.testutils import hybrid_cloud
-    from sentry.testutils.silo import assume_test_silo_mode
 
     # Exempt get_connection call from silo validation checks
-    with assume_test_silo_mode(SiloMode.MONOLITH):
+    with SiloMode.exit_single_process_silo_context(), SiloMode.enter_single_process_silo_context(
```
Comment (corps) [inline src/sentry/db/postgres/transactions.py:38]: This also addresses an issue where the `django_test_transaction_water_mark` would not work inside of an RPC call due to the way... override_settings works.  Fortunately, the `SiloMode.enter_single_process_silo_context` addresses this.

**Review by RyanSkonnord — APPROVED** [review]
It sucks that our setup/override code is getting so complex, but in terms of necessity we can't argue with the test behavior.

File: src/sentry/testutils/silo.py [hunk 2]
```diff
@@ -92,34 +92,51 @@ def _add_siloed_test_classes_to_module(
...
                     "silo_mode": silo_mode,
                     "regions": tuple(regions or _DEFAULT_TEST_REGIONS),
                     "is_acceptance_test": is_acceptance_test,
-                    "run": run,
+                    "_callSetUp": decorate_with_context(test_class._callSetUp),
+                    "_callTestMethod": decorate_with_context(test_class._callTestMethod),
                 },
             )
 
+            if hasattr(test_class, "_overridden_settings") and test_class._overridden_settings:
```
Comment (markstory) [inline src/sentry/testutils/silo.py:135]: What sets `_overridden_settings`? I don't see any references to it in master, and this is the only usage in this diff too.

Comment (corps) [reply src/sentry/testutils/silo.py:135]: It's how `override_settings` works when applied to classes.  Actually, this isn't the bug, I've been digging and I think it's even more subtle....

File: src/sentry/testutils/silo.py [hunk 3]
```diff
@@ -87,37 +88,49 @@ def _is_acceptance_test(test_class: type) -> bool:
...
-                name,
-                (test_class,),
-                {
-                    "silo_mode": silo_mode,
-                    "regions": tuple(regions or _DEFAULT_TEST_REGIONS),
-                    "is_acceptance_test": is_acceptance_test,
-                    "run": run,
-                },
+        final_regions = tuple(regions or _DEFAULT_TEST_REGIONS)
+        # settings: Dict[str, Any] | None = getattr(test_class, '_overridden_settings', None)
```
Comment (markstory) [inline src/sentry/testutils/silo.py:95]:
```suggestion
```

File: src/sentry/web/frontend/auth_login.py [hunk 4]
```diff
@@ -197,7 +197,7 @@ def post(self, request: Request, **kwargs) -> HttpResponse:
         if op == "sso" and request.POST.get("organization"):
             return self.redirect_post_to_sso(request=request)
 
-        organization = kwargs.pop("organization", None)
+        organization: RpcOrganization = kwargs.pop("organization", None)
```
Comment (markstory) [inline src/sentry/web/frontend/auth_login.py:200]:
```suggestion
        organization: RpcOrganization | None = kwargs.pop("organization", None)
```

Couldn't this be none as well?

File: tests/sentry/web/frontend/test_auth_login.py [hunk 5]
```diff
@@ -10,27 +10,38 @@
...
     @cached_property
     def path(self):
         return reverse("sentry-login")
 
+    def setUp(self) -> None:
+        from sentry.utils import redis
+
+        # Unfortunately, ratelimiting, which is backed by redis, will fail this test when it runs
+        # twice in silo mode, so we're forced to clear the redis cluster to block the rate limiting.
+        redis.redis_clusters.get("default").flushall(asynchronous=False)
```
Comment (markstory) [inline tests/sentry/web/frontend/test_auth_login.py:39]: If we end up needing this in more endpoint tests we can extract to conftest. Adding overhead to all tests just to resolve this problem seems heavy handed.

File: tests/sentry/web/frontend/test_auth_login.py [hunk 6]
```diff
@@ -222,15 +232,20 @@ def test_registration_single_org(self):
         user = User.objects.get(username="test-a-really-long-email-address@example.com")
 
         # User is part of the default org
-        default_org = Organization.get_default()
-        org_member = OrganizationMember.objects.get(organization_id=default_org.id, user_id=user.id)
+        with assume_test_silo_mode(SiloMode.REGION):
+            default_org = Organization.get_default()
+        with assume_test_silo_mode(SiloMode.REGION):
+            org_member = OrganizationMember.objects.get(
```
Comment (markstory) [inline tests/sentry/web/frontend/test_auth_login.py:238]:
```suggestion
        with assume_test_silo_mode(SiloMode.REGION):
            default_org = Organization.get_default()
            org_member = OrganizationMember.objects.get(
```

Should we combine these?

**Review by markstory — APPROVED** [review]

File: tests/sentry/web/frontend/test_auth_login.py (hunk 5 above)
Comment (corps) [reply tests/sentry/web/frontend/test_auth_login.py:39]: I agree

**Pull request merged**